scripts/flights2.py

```diff
@@ -0,0 +1,727 @@
+
+type IntoDate = dt.date | dt.datetime | YearMonthDay
+"""Anything that can be converted into a ``datetime.date``."""
+
+type IntoDateRange = (
+    tuple[IntoDate, IntoDate] | Mapping[Literal["start", "end"], IntoDate]
+)
+"""Anything that can be converted into a ``DateRange``."""
+
+
+type PlScanCsv = (
+    str
+    | Path
+    | IO[str]
+    | IO[bytes]
+    | bytes
+    | list[str]
+    | list[Path]
+    | list[IO[str]]
+    | list[IO[bytes]]
+    | list[bytes]
+)
+
+
+class DateTimeFormat(StrEnum):
+    ISO = "iso"
+    DECIMAL = "decimal"
+    ACTUALLY_ISO = "ISO"
+
+
+DTF_TO_FMT: Mapping[DateTimeFormat, LiteralString] = {
+    DateTimeFormat.ISO: "%Y/%m/%d %H:%M",
+    DateTimeFormat.ACTUALLY_ISO: "iso",
+}
+
+BASE_URL: LiteralString = "https://www.transtats.bts.gov/"
+ROUTE_ZIP: LiteralString = f"{BASE_URL}PREZIP/"
+REPORTING_PREFIX: LiteralString = (
+    "On_Time_Reporting_Carrier_On_Time_Performance_1987_present_"
+)
+ZIP: Literal[".zip"] = ".zip"
+GZIP: Literal[".csv.gz"] = ".csv.gz"
+PATTERN_GZIP: LiteralString = f"*{REPORTING_PREFIX}*{GZIP}"
+
+COLUMNS_DEFAULT: Sequence[Column] = (
+    "date",
+    "delay",
+    "distance",
+    "origin",
+    "destination",
+)
+"""Copied default from ``flights.py``."""
+
+SCAN_SCHEMA: pl.Schema = pl.Schema({
+    "FlightDate": pl.Date,
+    "CRSDepTime": pl.String,
+    "DepTime": pl.String,
+    "DepDelay": pl.Float64,
+    "ArrDelay": pl.Float64,
+    "Distance": pl.Float64,
+    "Origin": pl.String,
+    "Dest": pl.String,
+    "Cancelled": pl.Float64,
+})
+
+
+def _approx_latest(*, months_ago: int) -> dt.date:
+    # Very loose, aiming for the last day of `today - months_ago`
+    # In December, months_ago = 3 -> (2024, 8, 31)
+    weeks_ago = dt.timedelta(weeks=months_ago * 4)
+    current_month_start = dt.date.today().replace(day=1)
+    return (current_month_start - weeks_ago).replace(day=1) - dt.timedelta(days=1)
+
+
+def _into_date(obj: IntoDate, /) -> dt.date:
+    """Normalize date input."""
+    if isinstance(obj, dt.datetime):
+        return obj.date()
+    if isinstance(obj, dt.date):
+        return obj
+    if isinstance(obj, Sequence):
+        match obj:
+            case int(year), int(month), int(day):
+                return dt.date(year, month, day)
+            case int(year), int(month):
+                return dt.date(year, month, 1)
+            case (int(year),):
+                return dt.date(year, 1, 1)
+            case _:
+                raise TypeError(type(obj))
+    else:
+        raise TypeError(type(obj))
+
+
+class DateRange:
+    """
+    Matching a time period w/ required files.
+
+    - Validates provided dates are in range of known data
+    - Converts (start, end) to monthly file names
+    - Acts as a key, for detecting unique periods
+
+    Notes
+    -----
+    - Latest Available Data: August 2024 (in December 2024)
+        - 2024-12-07 -> max 2024-08-31
+    - https://www.transtats.bts.gov/releaseinfo.asp
+    """
```

Comment:
### TODO
- [ ] Clean up doc

Reply:
Resolved in (https://github.com/vega/vega-datasets/pull/645/commits/2d47fdd162c5d77dab68b7a7033609c7fe329da4)